### DISCUSSION TITLE
Create / Update multiple records at one time

### DISCUSSION THREAD

**Discussion by hungcrush** [pocketbase/pocketbase#1842] [Q&A] [answered]
Hi~
Could you give me some ideas on how to create an API that can both update existing records and create new ones if they do not exist, for multiple records?

Thank you so much 😬

**Comment by ganigeorgiev** [comment]
The question is too broad and I'm not sure what to recommend.

It depends on your use case and it can get complicated if:
- the records are from different collections
- you want to check the create/update collection(s) API rules for each individual record
- you want to support both `json` and `multipart/form-data` requests to allow uploading files

_You can also check some of the notes in https://github.com/pocketbase/pocketbase/issues/48#issuecomment-1211775627._


Could you elaborate a little more on your use case and what are you trying to do?

**Reply by hungcrush** [reply]
@ganigeorgiev 

I am creating a settings module. My settings are stored as Key-Value pairs like this:

![CleanShot 2023-02-12 at 23 40 13](https://user-images.githubusercontent.com/8098916/218324293-cd0a03c6-5b48-41eb-b874-412743bedfac.jpg)

in user UI:

![CleanShot 2023-02-12 at 23 45 04](https://user-images.githubusercontent.com/8098916/218324519-eafce304-52d6-4882-8976-f329eb9c0797.jpg)

So, when a user presses the submit button, it will send all the settings to the API. I have to check whether to create or update these keys.

**Comment by ganigeorgiev** [comment, marked as answer]
For this particular use case, I think the simplest solution will be to change your collection fields to one "settings" `json` field and create a single a record. Then you can just update it as usual, while creating/updating/deleting any key you want from the json:
```js
await pb.collection("settings").update("RECORD_ID", {
    settings: {
        "site_title": 123,
        "site_description": "test",
    },
})
```

Alternatively, if you want to maintain your current collections structure, you can create a new endpoint that will read a key-value object from the response and will "upsert" the settings within a transaction. Something like:

```go
package main

import (
	"log"
	"net/http"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/daos"
	"github.com/pocketbase/pocketbase/models"
)

func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.AddRoute(echo.Route{
			Method: http.MethodPost,
			Path:   "/upsert-settings",
			Handler: func(c echo.Context) error {
				data := map[string]any{}

				// load request data
				if err := c.Bind(&data); err != nil {
					return err
				}

				settingsCollection, err := app.Dao().FindCollectionByNameOrId("settings")
				if err != nil {
					return apis.NewBadRequestError("Failed to upsert settings", err)
				}

				app.Dao().RunInTransaction(func(txDao *daos.Dao) error {
					for k, v := range data {
						record, _ := txDao.FindFirstRecordByData("settings", "key", k)

						// create if missing
						if record == nil {
							record = models.NewRecord(settingsCollection)
							record.Set("key", k)
						}

						record.Set("value", v)

						if err := txDao.Save(record); err != nil {
							return apis.NewBadRequestError("Failed to upsert setting "+k, err)
						}
					}

					return nil
				})

				return c.NoContent(http.StatusNoContent)
			},
			Middlewares: []echo.MiddlewareFunc{
				apis.ActivityLogger(app),
				// add other middleware to guard the request if necessary...
				// you can check https://pocketbase.io/docs/router/#middlewares
			},
		})

		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

**Reply by hungcrush** [reply]
Thank you very much.... i got a lot of ideas from your answer. I will try with `json` field first.
